File: android/app/src/main/java/com/cordbot/maker/BotService.java
package com.cordbot.maker;

import com.cordbot.core.CordCore;
import android.app.*;
import android.content.Intent;
import android.os.*;
import androidx.core.app.NotificationCompat;
import net.dv8tion.jda.api.JDABuilder;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.requests.GatewayIntent;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import java.util.Map;

public class BotService extends Service {
    private JDA jda;
    private PowerManager.WakeLock wakeLock;

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        String code = intent.getStringExtra("CODE");
        String token = intent.getStringExtra("TOKEN");
        boolean useSlash = CordCore.getSetting(code, "slash", "ano").equalsIgnoreCase("ano");

        NotificationChannel channel = new NotificationChannel("BotChannel", "CordBot Status", NotificationManager.IMPORTANCE_LOW);
        getSystemService(NotificationManager.class).createNotificationChannel(channel);
        Notification notification = new NotificationCompat.Builder(this, "BotChannel")
            .setContentTitle("Bot je aktivní 24/7")
            .setContentText("Kód běží perfektně.")
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .build();
        startForeground(1, notification);

        PowerManager pm = (PowerManager) getSystemService(POWER_SERVICE);
        wakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "CordBot::Lock");
        wakeLock.acquire();

        new Thread(() -> {
            try {
                CordScriptParser parser = new CordScriptParser(code);
                jda = JDABuilder.createDefault(token)
                    .enableIntents(GatewayIntent.MESSAGE_CONTENT)
                    .addEventListeners(parser)
                    .build();
                jda.awaitReady();

                // AUTOMATICKÁ REGISTRACE SLASH PŘÍKAZŮ
                if (useSlash) {
                    for (String cmdName : parser.getCommands().keySet()) {
                        jda.upsertCommand(cmdName, "CordBot automatický příkaz").queue();
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();

        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        if (jda != null) jda.shutdownNow();
        if (wakeLock != null && wakeLock.isHeld()) wakeLock.release();
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) { return null; }
}

class CordScriptParser extends ListenerAdapter {
    private String prefix;
    private Map<String, String> commands;

    public CordScriptParser(String code) {
        this.prefix = CordCore.getSetting(code, "prefix", "!");
        this.commands = CordCore.getCommands(code);
    }

    public Map<String, String> getCommands() { return commands; }

    @Override
    public void onMessageReceived(MessageReceivedEvent event) {
        if (event.getAuthor().isBot()) return;
        String msg = event.getMessage().getContentRaw();
        if (msg.startsWith(prefix)) {
            String cmd = msg.substring(prefix.length()).trim().split(" ")[0];
            if (commands.containsKey(cmd)) {
                event.getChannel().sendMessage(commands.get(cmd)).queue();
            }
        }
    }

    // OBLUHA SLASH PŘÍKAZŮ PRO ANDROID!
    @Override
    public void onSlashCommandInteraction(SlashCommandInteractionEvent event) {
        String cmd = event.getName();
        if (commands.containsKey(cmd)) {
            event.reply(commands.get(cmd)).queue();
        }
    }
}

File: android/app/src/main/java/com/cordbot/maker/BotService.kt
package com.cordbot.maker

import android.app.*
import android.content.Intent
import android.os.*
import androidx.core.app.NotificationCompat
import net.dv8tion.jda.api.JDABuilder
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.requests.GatewayIntent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import java.util.regex.Pattern

class BotService : Service() {
    private var jda: JDA? = null
    private var wakeLock: PowerManager.WakeLock? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val token = intent?.getStringExtra("TOKEN") ?: return START_NOT_STICKY
        val code = intent.getStringExtra("CODE") ?: ""

        // 1. Udržení procesu 24/7 přes notifikaci
        val channelId = "BotServiceChannel"
        val channel = NotificationChannel(channelId, "Bot Status", NotificationManager.IMPORTANCE_LOW)
        getSystemService(NotificationManager::class.java).createNotificationChannel(channel)
        val notification = NotificationCompat.Builder(this, channelId)
            .setContentTitle("CordBot Běží 24/7")
            .setContentText("Bot je aktivní na pozadí.")
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .build()
        startForeground(1, notification)

        // 2. WakeLock (nedovolí CPU usnout)
        val powerManager = getSystemService(POWER_SERVICE) as PowerManager
        wakeLock = powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "CordBot::WakeLock")
        wakeLock?.acquire()

        // 3. Spuštění JDA (Nativní Discord Bot)
        Thread {
            try {
                jda = JDABuilder.createDefault(token)
                    .enableIntents(GatewayIntent.MESSAGE_CONTENT)
                    .addEventListeners(CordScriptParser(code))
                    .build()
            } catch (e: Exception) { e.printStackTrace() }
        }.start()

        return START_STICKY
    }

    override fun onDestroy() {
        jda?.shutdownNow()
        wakeLock?.let { if (it.isHeld) it.release() }
        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? = null
}

// Nativní CordScript Parser pro Kotlin
class CordScriptParser(private val code: String) : ListenerAdapter() {
    private var prefix = "!"
    private val commands = mutableMapOf<String, String>()

    init {
        val prefixMatcher = Pattern.compile("prefix:\\s*\"([^\"]+)\"").matcher(code)
        if (prefixMatcher.find()) prefix = prefixMatcher.group(1)!!

        val cmdMatcher = Pattern.compile("command\\s+(\\w+)\\s*\\{\\s*reply\\s+\"([^\"]+)\"\\s*\\}").matcher(code)
        while (cmdMatcher.find()) {
            commands[cmdMatcher.group(1)!!] = cmdMatcher.group(2)!!
        }
    }

    override fun onMessageReceived(event: MessageReceivedEvent) {
        if (event.author.isBot) return
        val msg = event.message.contentRaw
        if (msg.startsWith(prefix)) {
            val cmd = msg.removePrefix(prefix).trim().split(" ")[0]
            commands[cmd]?.let { reply ->
                event.channel.sendMessage(reply).queue()
            }
        }
    }
}
